src/adapter/src/coord/introspection.rs

```diff
@@ -95,16 +96,20 @@ pub fn check_cluster_restrictions(
     }
 
     // Only continue, and check restrictions, if a Plan would run some computation on the cluster.
-    match plan {
-        Plan::Subscribe(_) | Plan::Peek(_) | Plan::ReadThenWrite(_) => (),
+    //
+    // Note: Creating other objects like Materialized Views is prevented elsewhere. We define the
+    // 'mz_introspection' cluster to be "read-only", which restricts these actions.
+    let depends_on: Box<dyn Iterator<Item = GlobalId>> = match plan {
+        Plan::Subscribe(_) | Plan::ReadThenWrite(_) => Box::new(depends_on.into_iter().copied()),
+        // Some statements e.g. 'SHOW' ultimately depend on only system objects, _after_ planning.
```

Comment:
Oy, that's so annoying. I honestly think this is a bug/oversight. If planning adds `GlobalId`s to the query that weren't originally there, then we should be adding them to the `depends_on` variable.

Reply:
It's not so much that things get added to `depends_on` it's that what we depend on entirely changes. For example `SHOW INDEXES on <user_object>` initially describes it's dependencies as just `<user_object>`, which makes sense because it seems like we're generating `depends_on` solely from the [names in the statement](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/coord/command_handler.rs#L469). 

Only after planning do we truly know what the query will depend on. So maybe we shouldn't pass `depends_on` into this method at all 🤔

Reply:
I updated the code to not receive `depends_on` at all, and instead we get the dependencies from the Plans themselves. In the case of `ReadThenWrite` we only check the dependencies that we're reading from, which seemed correct? Because I don't think we care about inserts here, necessarily. More than happy to change this though :)

Reply:
I think I might disagree with you there. I think that `SHOW INDEXES on <user_object>` should be considered an alias for `SELECT ... FROM mz_indexes WHERE ...` and therefore should depend on not only `<user_object> but also `mz_indexes`.

`depends_on` isn't necessarily all of the names typed by the user, but all of the items that the statement depends on.

For subscribes we use the `depends_on` to generate a list of potential indexes to use, as well as to pick a timestamp. So it should contain `depends_on`.

For creating materialized views, `depends_on` is used for validations and generating an initial `since`, so if the materialized view is on a `SHOW INDEXES` command (if that's possible) then it should contain `mz_indexes`.

In fact, I think this might be the root cause of https://github.com/MaterializeInc/materialize/issues/18908.

Reply:
> I updated the code to not receive depends_on at all, and instead we get the dependencies from the Plans themselves. In the case of ReadThenWrite we only check the dependencies that we're reading from, which seemed correct? Because I don't think we care about inserts here, necessarily. More than happy to change this though :)

This approach works, but I am convinced that what you found is a bug an not the expected behavior.